src/core/pointcloud/qgspointcloudlayerexporter.cpp

```diff
@@ -0,0 +1,251 @@
+  const qint64 pointsToExport = std::max< qint64 >( std::min( mPointsLimit, pointCount ), 1 );
+  QgsPointCloudRequest request;
+  request.setAttributes( requestedAttributeCollection() );
+  std::unique_ptr<QgsPointCloudBlock> block = nullptr;
+  QgsFeatureList fl;
+  qint64 pointsExported = 0;
+  for ( const auto &node : nodes )
+  {
+    block.reset( index->nodeData( node, request ) );
+    const QgsPointCloudAttributeCollection attributesCollection = block->attributes();
+    const char *ptr = block->data();
+    int count = block->pointCount();
+    int recordSize = attributesCollection.pointRecordSize();
+    const QgsVector3D scale = block->scale();
+    const QgsVector3D offset = block->offset();
+    int xOffset;
+    int yOffset;
+    int zOffset;
+    attributesCollection.find( QStringLiteral( "X" ), xOffset );
+    attributesCollection.find( QStringLiteral( "Y" ), yOffset );
+    attributesCollection.find( QStringLiteral( "Z" ), zOffset );
+    for ( int i = 0; i < count; ++i )
+    {
+
+      if ( mFeedback )
+      {
+        mFeedback->setProgress( pointsExported / pointsToExport );
+        if ( mFeedback->isCanceled() )
+          return;
+      }
+
+      if ( pointsExported > mPointsLimit )
+        break;
+
+      double x, y, z;
+      QgsPointCloudAttribute::getPointXYZ( ptr, i, recordSize,
+                                           xOffset, QgsPointCloudAttribute::DataType::Int32,
+                                           yOffset, QgsPointCloudAttribute::DataType::Int32,
+                                           zOffset, QgsPointCloudAttribute::DataType::Int32,
+                                           scale, offset,
+                                           x, y, z );
+      if ( ! mZRange.contains( z ) ||
+           ! mExtent.contains( x, y ) )
+      {
+        ++pointsSkipped;
+        continue;
+      }
+
+      QgsFeature feature;
+      ct.transformInPlace( x, y, z );
+      feature.setGeometry( QgsGeometry( new QgsPoint( x, y, z ) ) );
+      const auto attributeMap = QgsPointCloudAttribute::getAttributeMap( ptr, i * recordSize, attributesCollection );
+      QgsAttributes featureAttributes;
+      for ( const QString &attribute : std::as_const( mRequestedAttributes ) )
+      {
+        const double val = attributeMap[ attribute ].toDouble();
+        featureAttributes.append( val );
+      }
+      feature.setAttributes( featureAttributes );
+      fl.append( feature );
+      ++pointsExported;
+    }
+  }
+  sink->addFeatures( fl );
+}
+
+QgsVectorLayer *QgsPointCloudLayerExporter::exportToMemoryLayer()
+{
+  if ( ! mLayer ||
+       ! mLayer->dataProvider() ||
+       ! mLayer->dataProvider()->index() ||
+       ! mLayer->dataProvider()->index()->isValid() )
+    return nullptr;
+
+  std::unique_ptr< QgsVectorLayer > layer( QgsMemoryProviderUtils::createMemoryLayer( mName, outputFields(), QgsWkbTypes::PointZ, mCrs ) );
+  layer->startEditing();
+  exportToSink( layer.get() );
+  layer->commitChanges();
```

Comment:
Don't use the vector layer buffer here -- write directly to layers->dataProvider() instead. It'll be significantly faster.